Do not use integer as keyword in weightsQIE11 PSet

JHiltbrand opened the pull request:
#### PR description:

In response to #32047, I have removed the use of just a simple integer (inside a string) when specifying the `weightsQIE11` PSet.
Although things currently work as is, using a bare integer breaks things when dumping out the config via `edmConfigDump`. I simply append string "ieta" to the bare integers and now the dumped out config also should work.

Commit:
do not use integer as keyword in weightsQIE11 PSet

Commit:
do string manipulation during loading of weights

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32067/19685

- This PR adds an extra 28KB to repository

cmsbuild commented:
A new Pull Request was created by  @JHiltbrand (Joshua Hiltbrand) for master.

It involves the following packages:

SimCalorimetry/HcalTrigPrimAlgos
SimCalorimetry/HcalTrigPrimProducers


@cmsbuild, @rekovic, @jmduarte can you please review it and eventually sign? Thanks.
@abdoulline, @rovere this is something you requested to watch as well.
@silviodonato, @dpiparo, @qliphy you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

silviodonato commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

silviodonato commented on SimCalorimetry/HcalTrigPrimProducers/python/hcaltpdigi_cfi.py:
```diff
@@ -17,34 +17,34 @@
-        ("25", cms.untracked.vdouble(1.0, 1.0)),
-        ("26", cms.untracked.vdouble(1.0, 1.0)),
-        ("27", cms.untracked.vdouble(1.0, 1.0)),
-        ("28", cms.untracked.vdouble(1.0, 1.0))
+        ("ieta1",  cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta2",  cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta3",  cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta4",  cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta5",  cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta6",  cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta7",  cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta8",  cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta9",  cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta10", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta11", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta12", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta13", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta14", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta15", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta16", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta17", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta18", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta19", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta20", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta21", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta22", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta23", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta24", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta25", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta26", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta27", cms.untracked.vdouble(1.0, 1.0)),
+        ("ieta28", cms.untracked.vdouble(1.0, 1.0))
         ])
```
```suggestion
    weightsQIE11 = cms.untracked.PSet(
        ieta1 =  cms.untracked.vdouble(1.0, 1.0),
        ieta2 =  cms.untracked.vdouble(1.0, 1.0),
        ieta3 =  cms.untracked.vdouble(1.0, 1.0),
        ieta4 =  cms.untracked.vdouble(1.0, 1.0),
        ieta5 =  cms.untracked.vdouble(1.0, 1.0),
        ieta6 =  cms.untracked.vdouble(1.0, 1.0),
        ieta7 =  cms.untracked.vdouble(1.0, 1.0),
        ieta8 =  cms.untracked.vdouble(1.0, 1.0),
        ieta9 =  cms.untracked.vdouble(1.0, 1.0),
        ieta10 = cms.untracked.vdouble(1.0, 1.0),
        ieta11 = cms.untracked.vdouble(1.0, 1.0),
        ieta12 = cms.untracked.vdouble(1.0, 1.0),
        ieta13 = cms.untracked.vdouble(1.0, 1.0),
        ieta14 = cms.untracked.vdouble(1.0, 1.0),
        ieta15 = cms.untracked.vdouble(1.0, 1.0),
        ieta16 = cms.untracked.vdouble(1.0, 1.0),
        ieta17 = cms.untracked.vdouble(1.0, 1.0),
        ieta18 = cms.untracked.vdouble(1.0, 1.0),
        ieta19 = cms.untracked.vdouble(1.0, 1.0),
        ieta20 = cms.untracked.vdouble(1.0, 1.0),
        ieta21 = cms.untracked.vdouble(1.0, 1.0),
        ieta22 = cms.untracked.vdouble(1.0, 1.0),
        ieta23 = cms.untracked.vdouble(1.0, 1.0),
        ieta24 = cms.untracked.vdouble(1.0, 1.0),
        ieta25 = cms.untracked.vdouble(1.0, 1.0),
        ieta26 = cms.untracked.vdouble(1.0, 1.0),
        ieta27 = cms.untracked.vdouble(1.0, 1.0),
        ieta28 = cms.untracked.vdouble(1.0, 1.0)
    )
```
What about this?

JHiltbrand replied:
Hmm, so the key must be a string, otherwise it gets interpreted as a variable.

silviodonato replied:
It should work exactly in the same way in both cases. 
I checked it with
```
import FWCore.ParameterSet.Config as cms

weightsQIE11 = cms.untracked.PSet(**dict([
        ("ieta1",  cms.untracked.vdouble(1.0, 1.0)),
        ("ieta2",  cms.untracked.vdouble(1.0, 1.0)),
        ("ieta3",  cms.untracked.vdouble(1.0, 1.0)),
        ]))

weightsQIE11_new = cms.untracked.PSet(
        ieta1 = cms.untracked.vdouble(1.0, 1.0),
        ieta2 = cms.untracked.vdouble(1.0, 1.0),
        ieta3 = cms.untracked.vdouble(1.0, 1.0),
        )


print(weightsQIE11.dumpPython())
print(weightsQIE11_new.dumpPython())
print()
print(weightsQIE11.ieta1)
print(weightsQIE11_new.ieta1)
print()
print(getattr(weightsQIE11,"ieta1"))
print(getattr(weightsQIE11_new,"ieta1"))
```

JHiltbrand replied:
Ah, I see. I will make this simplification.

JHiltbrand replied:
Addressed in 9877930.

silviodonato commented:
From https://github.com/cms-sw/cmssw/blob/master/SimCalorimetry/HcalTrigPrimProducers/src/HcalTrigPrimDigiProducer.cc#L48 I see that `weightsQIE11` parameter is not optional.
I would suggest to move it to a tracked parameter.

JHiltbrand commented:
> 
> 
> From https://github.com/cms-sw/cmssw/blob/master/SimCalorimetry/HcalTrigPrimProducers/src/HcalTrigPrimDigiProducer.cc#L48 I see that `weightsQIE11` parameter is not optional.
> I would suggest to move it to a tracked parameter.

Ah, you can do.

Commit:
change weightsQIE11 to tracked parameter

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32067/19706

- This PR adds an extra 28KB to repository

cmsbuild commented:
Pull request #32067 was updated. @cmsbuild, @rekovic, @jmduarte can you please check and sign again.

cmsbuild commented:
+1
Tested at: 00aba17901deb1c7dcdad55abe1a2e4e61cd7826
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b0570b/10594/summary.html
CMSSW: CMSSW_11_2_X_2020-11-09-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b0570b/10594/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2529296
 - DQMHistoTests: Total failures: 8
 - DQMHistoTests: Total nulls: 1
 - DQMHistoTests: Total successes: 2529265
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: -0.004 KiB( 34 files compared)
 - DQMHistoSizes: changed ( 312.0 ): -0.004 KiB       MessageLogger/Warnings
 - Checked 148 log files, 22 edm output root files, 35 DQM output files

silviodonato commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: 7ab883bf4f877f9c30a8eab43cf0371a136672ab
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b0570b/10606/summary.html
CMSSW: CMSSW_11_2_X_2020-11-09-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

VinInn commented on SimCalorimetry/HcalTrigPrimAlgos/interface/HcalTriggerPrimitiveAlgo.h:
```diff
@@ -124,7 +124,7 @@ class HcalTriggerPrimitiveAlgo {
   bool peakfind_;
   std::vector<double> weights_;
-  std::map<int, std::vector<double>> weightsQIE11_;
+  std::map<unsigned int, std::vector<double>> weightsQIE11_;
```
why a map?
would just std::array<std::array<double,2>,28> not do it?

JHiltbrand replied:
No profound reason to use a map, just out comfort. I have adopted the array of arrays approach.

VinInn commented on SimCalorimetry/HcalTrigPrimAlgos/src/HcalTriggerPrimitiveAlgo.cc:
```diff
@@ -858,7 +858,8 @@ void HcalTriggerPrimitiveAlgo::setWeightsQIE11(const edm::ParameterSet& weightsQ
   for (auto& ietaStr : ietaStrs) {
-    weightsQIE11_[std::stoi(ietaStr)] = weightsQIE11.getUntrackedParameter<std::vector<double>>(ietaStr);
+    // Strip off "ieta" part of key and just use integer value in map
+    weightsQIE11_[std::stoi(ietaStr.substr(4))] = weightsQIE11.getUntrackedParameter<std::vector<double>>(ietaStr);
```
this will then become
```
auto const & v = weightsQIE11.getUntrackedParameter<std::vector<double>>(ietaStr);
weightsQIE11_[std::stoi(ietaStr.substr(4))-1] = {v[0],v[1]};
```

btw WHY "Untracked"? are These parameters not changing the output of the simulation?

VinInn replied:
fi you want to keep the addressing between 1-28, just make it 29-long and waste 0...

JHiltbrand replied:
I removed the last bit of "untracked" from this. Also, I decided to make the array 29 long to avoid any `-1`.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b0570b/10606/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2529296
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2529273
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 148 log files, 22 edm output root files, 35 DQM output files

Commit:
deprecate map in favor of array of arrays

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32067/19722

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #32067 was updated. @cmsbuild, @rekovic, @jmduarte can you please check and sign again.

silviodonato commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

silviodonato commented:
@JHiltbrand please look at https://github.com/cms-sw/cmssw/pull/32067#discussion_r520387250

Commit:
address comment from Silvio

Commit:
small fix for when using non-default filters

JHiltbrand commented:
Ok, I believe I have addressed open concerns. I also added a small fix (1f286f966ed9fbb7485fd2184d2632d97f7d4f85) for when running with non-default filters (for future).

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-32067/19728

- This PR adds an extra 32KB to repository

cmsbuild commented:
Pull request #32067 was updated. @cmsbuild, @rekovic, @jmduarte can you please check and sign again.

cmsbuild commented:
+1
Tested at: 665c124f5f6fe7b9bd8f4f082820470f7f81f145
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b0570b/10618/summary.html
CMSSW: CMSSW_11_2_X_2020-11-10-1100
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b0570b/10618/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 4 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2529296
 - DQMHistoTests: Total failures: 7
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2529267
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 148 log files, 22 edm output root files, 35 DQM output files

silviodonato commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

silviodonato commented:
@cms-sw/l1-l2 do you have any comments? I'm going to include this in pre9

cmsbuild commented:
+1
Tested at: 1f286f966ed9fbb7485fd2184d2632d97f7d4f85
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b0570b/10638/summary.html
CMSSW: CMSSW_11_2_X_2020-11-10-2300
SCRAM_ARCH: slc7_amd64_gcc820

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-b0570b/10638/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 0 differences found in the comparisons
 - DQMHistoTests: Total files compared: 35
 - DQMHistoTests: Total histograms compared: 2529296
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 2529273
 - DQMHistoTests: Total skipped: 22
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 34 files compared)
 - Checked 148 log files, 22 edm output root files, 35 DQM output files

silviodonato commented:
merge
